scai-gen/cmd/sigstore.go

```diff
@@ -0,0 +1,129 @@
+// adapted from https://github.com/slsa-framework/slsa-github-generator/blob/main/signing/sigstore/fulcio.go
+// and https://github.com/slsa-framework/slsa-github-generator/blob/main/internal/builders/generic/attest.go
+package cmd
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+
+	"github.com/in-toto/scai-demos/scai-gen/fileio"
+
+	ita "github.com/in-toto/attestation/go/v1"
+	"github.com/sigstore/cosign/v2/cmd/cosign/cli/fulcio"
+	"github.com/sigstore/cosign/v2/cmd/cosign/cli/options"
+	"github.com/sigstore/cosign/v2/cmd/cosign/cli/sign"
+	"github.com/sigstore/cosign/v2/pkg/providers"
+	"github.com/sigstore/sigstore/pkg/signature/dsse"
+	"github.com/slsa-framework/slsa-github-generator/signing/envelope"
+	"github.com/slsa-framework/slsa-github-generator/signing/sigstore"
+	"github.com/spf13/cobra"
+	"google.golang.org/protobuf/encoding/protojson"
+)
+
+var sigstoreCmd = &cobra.Command{
+	Use:   "sigstore",
+	Args:  cobra.ExactArgs(1),
+	Short: "Use Sigstore signing to create a signed DSSE for an in-toto Statements",
+	RunE:  signWithSigstore,
+}
+
+// attestation is a signed attestation.
+type attestation struct {
+	cert []byte
+	att  []byte
+}
+
+// Bytes returns the signed attestation as an encoded DSSE JSON envelope.
+func (a *attestation) Bytes() []byte {
+	return a.att
+}
+
+// Cert returns the certificate used to sign the attestation.
+func (a *attestation) Cert() []byte {
+	return a.cert
+}
+
+func init() {
+	sigstoreCmd.Flags().StringVarP(
+		&outFile,
+		"out-file",
+		"o",
+		"",
+		"Filename to write out the signed DSSE object",
+	)
+	sigstoreCmd.MarkFlagRequired("out-file") //nolint:errcheck
+}
+
+func getNewFulcioSigner(ctx context.Context) (*fulcio.Signer, error) {
+	ko := options.KeyOpts{
+		OIDCIssuer:   options.DefaultOIDCIssuerURL,
+		OIDCClientID: "sigstore",
+		FulcioURL:    options.DefaultFulcioURL,
+	}
+
+	sv, err := sign.SignerFromKeyOpts(ctx, "", "", ko)
+	if err != nil {
+		return nil, fmt.Errorf("getting Fulcio signer: %w", err)
+	}
+
+	return fulcio.NewSigner(ctx, ko, sv)
+}
+
+func signWithSigstore(_ *cobra.Command, args []string) error {
+	fmt.Println("EXPERIMENTAL FEATURE. DO NOT USE IN PRODUCTION.")
+
+	statementFile := args[0]
+	statement := &ita.Statement{}
+	err := fileio.ReadPbFromFile(statementFile, statement)
+	if err != nil {
+		return err
+	}
+
+	// will only sign valid Statements
+	err = statement.Validate()
+	if err != nil {
+		return fmt.Errorf("invalid in-toto Statement: %w", err)
+	}
+
+	ctx := context.Background()
+
+	// Get Fulcio signer
+	if !providers.Enabled(ctx) {
```

Comment:
I'd re-enable cmd in the args and use `cmd.Context()`. If one isn't set, it eventually hits `Background` too.

Reply:
Just to be 100% sure I understood the change you're suggesting. I should replace the `providers.Enabled(ctx)` check with `providers.Enabled(cmd.Context())?

Reply:
Yes: https://pkg.go.dev/github.com/spf13/cobra#Command.Context